lib/directory-view.coffee

```diff
@@ -108,10 +108,12 @@ class DirectoryView
       view = new FileView(entry)
 
     subscription = @directory.onDidRemoveEntries (removedEntries) ->
-      for removedName, removedEntry of removedEntries when entry is removedEntry
+      removedEntries.forEach (removedEntry, removedName) ->
+        return unless entry is removedEntry
         view.element.remove()
         subscription.dispose()
-        break
+        # TODO: When this file is converted to JS, convert this forEach loop
+        # to a for...of loop and add a break here for performance
```

Comment:
Another option is to avoid the "problem" by using `Array.prototype.some` instead;

``` coffee
removedEntries.some (removedEntry, removedName) ->
  return true unless entry is removedEntry
  view.element.remove()
  subscription.dispose()
``` 

This code will behave essentially exactly like the old version — it will iterate until the desired item is encountered, then short-circuit the iteration. You could of course also achieve this by using `Array.prototype.every`, but the version above more closely mimics the old version I think (at least in terms of lines of code, and overall legibility).

Reply:
Since you already checked that using `Map`s does not break consumers of this API, I wonder if we could change the parameters to make this code path more efficient. In particular, what about exposing a `Set` of entries that were removed so that we can retrieve it in `O(1)` here? Like:

```js
if (removedEntries.has(entry)) {
  // cleanup
}
```

An alternative might be to rely on the name and avoid changing the API altogether:

```js
if (removedEntries.has(entry.name)) {
  // cleanup
}
```

Although I am slightly concerned about the order of events when a rename occurs if we decide to go choose the second option.

What do you think? 💭

Reply:
> could you expand on your concerns about the event order with the second option?

Yeah, sorry if that comment was a little obscure. My primary concern is that in the same tick of the loop the following could happen:

1. An entry with name `x` is added.
2. A previous entry with name `x` is deleted. 

I am not sure this would be possible in the current model, but if we receive change events asynchronously I worry that the add and the delete event could get grouped together and cause the wrong entry (or both, in the example above) to be deleted.

If that's a circumstance that we are sure we will never encounter, then I think we could go ahead and implement 2).